### PR TITLE
Smoothing for Pseudotime kernel

### LINKED ISSUE #504: pseudotime-based kernel tutorial
- [ ] Additional function parameters, change functionality or change defaults?
- [ ] New estimator in `cr.tl.estimators`?
- [ ] New kernel in `cr.tl.kernels`?
- [ ] New gene trend model in `cr.ul.models`?
- [ ] New plotting function in `cr.pl`?
- [ ] Other?

This should be a tutorial that demonstrates use cases for soft/hard pseudotime kernels (i.e. Palantir/VIA) as well as for the other anticipated new kernel...

### PR DESCRIPTION
**Pull request by michalk8**
**IMPORTANT: Please search among the [Pull request](../pulls) before creating one.**

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] New feature (non-breaking change which adds functionality)

## Description

Adds soft/hard/custom threhdolding for cells in PT's past for `PseudotimeKernel`.
@Marius1311 would be great if you could check this out and compare it with VIA's implementation [here](https://github.com/ShobiStassen/VIA/blob/master/VIA/core.py#L635).
Briefly, I've kept all the constant the same, except for `nu` (now is 0.5, orig. was 1 for the gen. [logistic regression](https://en.wikipedia.org/wiki/Generalised_logistic_function)) + I do 95% winsorization, instead of 90% on the KNN weights.
You can set the thresholding as `threshold_scheme={'soft','hard'}` or pass a function which has the following signature:
`(cell_pseudotime: float, neigh_pseudotime: np.ndarray, neigh_dist: np.ndarray, **kwargs: Any) -> np.ndarray`
(not sure about the naming, `neigh_dist` are the `KNN` weights, should be renamed).

## How has this been tested?

No tests yet.

## Closes

closes #504

TODO:
- [x] tests
- [x] determine which parameters to expose
- [x] update docs (@Marius1311 I leave this up to you)
- [ ] asses the robustness of parameters (not really robust to perc. clipping [expected imho], not sure about others).

### CODE REVIEW

File: cellrank/tl/kernels/_pseudotime_schemes.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        neigh_dist[past_ixs] = np.clip(
+            neigh_dist[past_ixs],
+            np.percentile(neigh_dist[past_ixs], 100 - perc),
+            np.percentile(neigh_dist[past_ixs], perc),
+        )
+
+        weights = np.full_like(neigh_dist, fill_value=0.5)
+
+        dt = cell_pseudotime - neigh_pseudotime[past_ixs]
+        weights[past_ixs] = 1.0 / ((1.0 + np.exp(b * dt)) ** (1.0 / nu))
```
[the remark this answers is not in the document]
Comment (Marius1311) [reply cellrank/tl/kernels/_pseudotime_schemes.py:162]: They also seem to be using nu=1 in their implementation.

**Review by Marius1311 — CHANGES_REQUESTED** [review]
Very elegant solution @michalk8. I suggest
- removing the clipping
- removing the nu parameter (fixing it to 1)
- making sure weights are in [0, 1] 
I detail this in my comments. Once this is done, I will take care of the docstrings and test this on some data.

File: cellrank/tl/kernels/_pseudotime_schemes.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+class SoftThresholdScheme(PseudotimeScheme):
+    """TODO."""
+
+    @d.dedent
+    def __call__(
+        self,
+        cell_pseudotime: float,
+        neigh_pseudotime: np.ndarray,
+        neigh_dist: np.ndarray,
+        b: float = 20.0,
```
[the remark this answers is not in the document]
Comment (michalk8) [reply cellrank/tl/kernels/_pseudotime_schemes.py:133]: b=1 is for `"using the pseudotime calculated from lazy-jumping walk`"
b=20 is for `" using the refined MCMC Psuedotimes before calculating lineage likelihood paths"`

File: cellrank/tl/kernels/_pseudotime_schemes.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        ----------
+        %(pt_scheme.parameters)s
+
+        Returns
+        -------
+        %(pt_scheme.returns)s
+        """
+        past_ixs = np.where(neigh_pseudotime < cell_pseudotime)[0]
+        if not len(past_ixs):
+            return neigh_dist * 0.5
```
[the remark this answers is not in the document]
Comment (michalk8) [reply cellrank/tl/kernels/_pseudotime_schemes.py:150]: It does matter: `sigmoid(0) = 0.5` i.e. when the current point is the same (or very close) w.r.t. to PT.

File: cellrank/tl/kernels/_pseudotime_schemes.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+        # TODO: should all of them be clipped, or just the ones in the past?
+        neigh_dist[past_ixs] = np.clip(
+            neigh_dist[past_ixs],
+            np.percentile(neigh_dist[past_ixs], 100 - perc),
+            np.percentile(neigh_dist[past_ixs], perc),
+        )
```
[the remark this answers is not in the document]
Comment (michalk8) [reply cellrank/tl/kernels/_pseudotime_schemes.py:157]: I can send you some figures I have where I did a grid search so you can see for yourself if it's necessary.
I would say keep it (i.e. cells which are further from each other should have greater KNN distance between them, this
dampens these "outliers" if weights are not sufficiently small).

File: cellrank/tl/kernels/_pseudotime_schemes.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    """TODO."""
+
+    @d.dedent
+    def __call__(
+        self,
+        cell_pseudotime: float,
+        neigh_pseudotime: np.ndarray,
+        neigh_dist: np.ndarray,
+        b: float = 20.0,
+        nu: float = 0.5,
```
[the remark this answers is not in the document]
Comment (Marius1311) [reply cellrank/tl/kernels/_pseudotime_schemes.py:134]: Hi Mike, please check your mattermost, I send you exactly these plots yesterday already and made the suggestion based on that.

**Comment by Marius1311** [comment]
What's the status here, only tests missing?

**Comment by michalk8** [comment]
> What's the status here, only tests missing?

From my side, it's the tests, from your side, it would be to polish the docstrings of `PseudotimeKernel` + of the schemes (I've alreay included the VIA citation)

**Comment by Marius1311** [comment]
Nice, will tend to this after my PR today!

**Comment by Marius1311** [comment]
Estimators don't like the pseudotime kernel yet, running
```
from cellrank.tl.kernels import PseudotimeKernel
from cellrank.tl.estimators import GPCCA

kernel = PseudotimeKernel(adata, time_key='gcs_pseudotime', check_connectivity=True)
kernel.compute_transition_matrix(threshold_scheme='hard', k=3)

g_fwd = GPCCA(kernel)
```
raises
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-38-a0174fcf3fc7> in <module>
----> 1 g_fwd = GPCCA(kernel)
      2 print(g_fwd)

~/Projects/cellrank/cellrank/tl/estimators/_base_estimator.py in __init__(self, obj, inplace, read_from_adata, obsp_key, g2m_key, s_key, write_to_adata, key)
    103         from anndata import AnnData
    104 
--> 105         super().__init__(obj, obsp_key=obsp_key, key=key, write_to_adata=write_to_adata)
    106 
    107         if isinstance(obj, (KernelExpression, AnnData)) and not inplace:

~/Projects/cellrank/cellrank/tl/estimators/_property.py in __init__(self, *args, **kwargs)
    864 
    865     def __init__(self, *args, **kwargs):
--> 866         super().__init__(*args, **kwargs)
    867 
    868         self._is_irreducible = None  # no need to refactor these

~/Projects/cellrank/cellrank/tl/estimators/_property.py in __init__(self, obj, key, obsp_key, write_to_adata)
    213             self._kernel = PrecomputedKernel(obj.obsp[obsp_key], adata=obj)
    214         else:
--> 215             raise TypeError(
    216                 f"Expected an object of type `KernelExpression`, `numpy.ndarray`, `scipy.sparse.spmatrix` "
    217                 f"or `anndata.AnnData`, got `{type(obj).__name__!r}`."

TypeError: Expected an object of type `KernelExpression`, `numpy.ndarray`, `scipy.sparse.spmatrix` or `anndata.AnnData`, got `'PseudotimeKernel'`.
```

**Comment by michalk8** [comment]
> Estimators don't like the pseudotime kernel yet, running

This is an issue with jupyter + autoreload (can't reproduce this).

**Comment by Marius1311** [comment]
Okay sorry, my mistake then!

**Comment by Marius1311** [comment]
Okay, you're right, restarting my kernel solved this, really sorry!

**Comment by Marius1311** [comment]
The docstrings should be done @michalk8 - let me know if I missed anyting!

**Comment by Marius1311** [comment]
Thanks for adding the tests!

**Pull request merged**